File: guests/go/dirty/dirty.go
// The dirty-state scene (tools/scenes/dirty.steps). TWO DECLARATIONS, on
// purpose: kaya does not watch your signals and guess.
package dirty

import (
	kaya "dev.kaya/bindings/go"
)

func App() *kaya.App {
	app := kaya.NewApp()

	var doc, status kaya.Signal[string]
	app.Build(func(tx *kaya.Tx) {
		// Dirty is NOT declared here: the script reads the clean window.
		win := tx.Window(0).Title("dirty").VetoClose(true)

		doc = tx.Signal("notes")
		status = tx.Signal("saved")

		win.OnCloseRequested(func(tx *kaya.Tx) {
			tx.ShowAlert().
				Title("unsaved changes").
				Message("the document has unsaved changes").
				Action("Discard").
				Cancel("Keep Editing").
				OnResult(func(tx *kaya.Tx, choice uint32) {
					if choice == kaya.AlertChoiceCancel {
						tx.Write(status, "kept editing")
						return
					}
					// ABORTS if it runs — docs/traps.md, "An app can VETO a
					// close but cannot AGREE to one".
					tx.DestroyWindow(0)
				}).
				Show()
		})

		tx.Mount(tx.Column(func() {
			tx.Label(doc)    // label#0
			tx.Label(status) // label#1
			tx.Button("edit", func(tx *kaya.Tx) { // button#0
				tx.Write(doc, "notes and a line")
				tx.Write(status, "unsaved")
				tx.Window(0).Dirty(true)
			})
			tx.Button("save", func(tx *kaya.Tx) { // button#1
				tx.Write(status, "saved")
				tx.Window(0).Dirty(false)
			})
		}))
	})

	return app
}

File: guests/java/dev/kaya/guests/Dirty.java
package dev.kaya.guests;

import dev.kaya.KayaApp;
import dev.kaya.KayaWire;

/**
 * The dirty scene from the JVM — guests/rust/dirty.rs, tools/scenes/dirty.steps.
 */
public final class Dirty {
    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            KayaApp.WindowRef win = tx.window(0).title("dirty").vetoClose(true);
            KayaApp.Signal<String> doc = tx.signal("notes");
            KayaApp.Signal<String> status = tx.signal("saved");

            win.onCloseRequested(t -> {
                t.showAlert()
                        .title("unsaved changes")
                        .message("the document has unsaved changes")
                        .action("Discard")
                        .cancel("Keep Editing")
                        .onResult((t2, choice) -> {
                            if (choice == KayaWire.ALERT_CHOICE_CANCEL) {
                                t2.write(status, "kept editing");
                            } else {
                                // Aborts if it ever runs (docs/traps.md, an app
                                // can VETO a close but cannot AGREE to one).
                                t2.destroyWindow(0);
                            }
                        })
                        .show();
            });

            tx.mount(tx.column(() -> {
                tx.label(doc); // label#0
                tx.label(status); // label#1
                // The document and the mark are two statements.
                tx.button("edit", t -> { // button#0
                    t.write(doc, "notes and a line");
                    t.write(status, "unsaved");
                    t.window(0).dirty(true);
                });
                tx.button("save", t -> { // button#1
                    t.write(status, "saved");
                    t.window(0).dirty(false);
                });
            }));
        });

        app.dispatchLoop();
    }

    private Dirty() {}
}
